Default to current_thread tokio runtime

wfchandler opened the pull request:
The vast majority operations in the CLI perform a single task, and we suspect that they do not benefit from the scalability offered by the multi-threaded runtime tokio defaults to.

Benchmarking the runtimes, `oxide version` unsurprisingly shows a significant improvement in performance with `current_thread`:

```
  $ hyperfine -L oxide './oxide-st, ./oxide-mt' -w3 '{oxide} version'
  Benchmark 1: ./oxide-st version
    Time (mean ± σ):      14.5 ms ±   0.6 ms    [User: 9.2 ms, System: 5.3 ms]
    Range (min … max):    13.3 ms …  17.9 ms    187 runs

  Benchmark 2:  ./oxide-mt version
    Time (mean ± σ):      22.1 ms ±   1.0 ms    [User: 11.6 ms, System: 13.3 ms]
    Range (min … max):    20.2 ms …  26.4 ms    127 runs

  Summary
    ./oxide-st version ran
      1.52 ± 0.10 times faster than  ./oxide-mt version
```

Of more relevance to actual use, a simple command that hits the network shows a smaller benefit to using `current_thread`:

```
  $ hyperfine -L oxide './oxide-st,./oxide-mt' '{oxide} instance list --project=will'
  Benchmark 1: ./oxide-st instance list --project=will
    Time (mean ± σ):     156.4 ms ±   5.8 ms    [User: 12.8 ms, System: 7.8 ms]
    Range (min … max):   149.4 ms … 174.7 ms    18 runs

  Benchmark 2:  ./oxide-mt instance list --project=will
    Time (mean ± σ):     163.6 ms ±   8.6 ms    [User: 14.9 ms, System: 17.5 ms]
    Range (min … max):   153.5 ms … 191.6 ms    18 runs

  Summary
    ./oxide-st instance list --project=will ran
      1.05 ± 0.07 times faster than  ./oxide-mt instance list --project=will
```

However, this mostly washes out when run from my more distant home network, though CPU usage remains slightly lower:

```
  $ hyperfine -M5 -L oxide './oxide-st,./oxide-mt' '{oxide} instance list --project=will'
  Benchmark 1: ./oxide-st instance list --project=will
    Time (mean ± σ):     541.2 ms ±  15.3 ms    [User: 13.9 ms, System: 13.8 ms]
    Range (min … max):   515.1 ms … 554.2 ms    5 runs

  Benchmark 2: ./oxide-mt instance list --project=will
    Time (mean ± σ):     544.5 ms ±   3.4 ms    [User: 16.4 ms, System: 19.7 ms]
    Range (min … max):   541.0 ms … 549.6 ms    5 runs

  Summary
    ./oxide-st instance list --project=will ran
      1.01 ± 0.03 times faster than ./oxide-mt instance list --project=will
```

With disk imports, which are highly parallel, we see higher variance with `current_thread`, but in human terms the results are essentially the same:

```
  $ hyperfine -L oxide './oxide-st,./oxide-mt' 'bash -c "{oxide} disk import --project will --description BenchRTs --path ~/alpine.raw --disk bench-$(date +%s%N) --disk-block-size 512"'
  Benchmark 1: bash -c "./oxide-st disk import --project will --description BenchRTs --path ~/alpine.raw --disk bench-$(date +%s%N) --disk-block-size 512"
    Time (mean ± σ):      6.817 s ±  1.341 s    [User: 0.545 s, System: 1.144 s]
    Range (min … max):    5.196 s …  8.928 s    10 runs

  Benchmark 2: bash -c "./oxide-mt disk import --project will --description BenchRTs --path ~/alpine.raw --disk bench-$(date +%s%N) --disk-block-size 512"
    Time (mean ± σ):      6.375 s ±  0.982 s    [User: 0.615 s, System: 1.113 s]
    Range (min … max):    4.975 s …  8.027 s    10 runs

  Summary
    bash -c "./oxide-mt disk import --project will --description BenchRTs --path ~/alpine.raw --disk bench-$(date +%s%N) --disk-block-size 512" ran
      1.07 ± 0.27 times faster than bash -c "./oxide-st disk import --project will --description BenchRTs --path ~/alpine.raw --disk bench-$(date +%s%N) --disk-block-size 512"
```

Update the CLI to use the `current_thread` runtime.

While we're at it, replace a few unwraps with errors.

wfchandler committed:
Default to current_thread tokio runtime

The vast majority operations in the CLI perform a single task, and we
suspect that they do not benefit from the scalability offered by the
multi-threaded runtime tokio defaults to.

Benchmarking the runtimes, `oxide version` unsurprisingly shows a
significant improvement in performance with `current_thread`:

  $ hyperfine -L oxide './oxide-st, ./oxide-mt' -w3 '{oxide} version'
  Benchmark 1: ./oxide-st version
    Time (mean ± σ):      14.5 ms ±   0.6 ms    [User: 9.2 ms, System: 5.3 ms]
    Range (min … max):    13.3 ms …  17.9 ms    187 runs

  Benchmark 2:  ./oxide-mt version
    Time (mean ± σ):      22.1 ms ±   1.0 ms    [User: 11.6 ms, System: 13.3 ms]
    Range (min … max):    20.2 ms …  26.4 ms    127 runs

  Summary
    ./oxide-st version ran
      1.52 ± 0.10 times faster than  ./oxide-mt version

Of more relevance to actual use, a simple command that hits the network
shows a smaller benefit to using `current_thread`:

  $ hyperfine -L oxide './oxide-st,./oxide-mt' '{oxide} instance list --project=will'
  Benchmark 1: ./oxide-st instance list --project=will
    Time (mean ± σ):     156.4 ms ±   5.8 ms    [User: 12.8 ms, System: 7.8 ms]
    Range (min … max):   149.4 ms … 174.7 ms    18 runs

  Benchmark 2:  ./oxide-mt instance list --project=will
    Time (mean ± σ):     163.6 ms ±   8.6 ms    [User: 14.9 ms, System: 17.5 ms]
    Range (min … max):   153.5 ms … 191.6 ms    18 runs

  Summary
    ./oxide-st instance list --project=will ran
      1.05 ± 0.07 times faster than  ./oxide-mt instance list --project=will

However, this mostly washes out when run from my more distant home
network, though CPU usage remains slightly lower:

  $ hyperfine -M5 -L oxide './oxide-st,./oxide-mt' '{oxide} instance list --project=will'
  Benchmark 1: ./oxide-st instance list --project=will
    Time (mean ± σ):     541.2 ms ±  15.3 ms    [User: 13.9 ms, System: 13.8 ms]
    Range (min … max):   515.1 ms … 554.2 ms    5 runs

  Benchmark 2: ./oxide-mt instance list --project=will
    Time (mean ± σ):     544.5 ms ±   3.4 ms    [User: 16.4 ms, System: 19.7 ms]
    Range (min … max):   541.0 ms … 549.6 ms    5 runs

  Summary
    ./oxide-st instance list --project=will ran
      1.01 ± 0.03 times faster than ./oxide-mt instance list --project=will

With disk imports, which are highly parallel, we see higher variance
with `current_thread`:

  $ hyperfine -L oxide './oxide-st,./oxide-mt' 'bash -c "{oxide} disk import --project will --description BenchRTs --path ~/alpine.raw --disk bench-$(date +%s%N) --disk-block-size 512"'
  Benchmark 1: bash -c "./oxide-st disk import --project will --description BenchRTs --path ~/alpine.raw --disk bench-$(date +%s%N) --disk-block-size 512"
    Time (mean ± σ):      6.817 s ±  1.341 s    [User: 0.545 s, System: 1.144 s]
    Range (min … max):    5.196 s …  8.928 s    10 runs

  Benchmark 2: bash -c "./oxide-mt disk import --project will --description BenchRTs --path ~/alpine.raw --disk bench-$(date +%s%N) --disk-block-size 512"
    Time (mean ± σ):      6.375 s ±  0.982 s    [User: 0.615 s, System: 1.113 s]
    Range (min … max):    4.975 s …  8.027 s    10 runs

  Summary
    bash -c "./oxide-mt disk import --project will --description BenchRTs --path ~/alpine.raw --disk bench-$(date +%s%N) --disk-block-size 512" ran
      1.07 ± 0.27 times faster than bash -c "./oxide-st disk import --project will --description BenchRTs --path ~/alpine.raw --disk bench-$(date +%s%N) --disk-block-size 512"

Update the CLI to use the `current_thread` runtime unless we are
performing a disk import. This requires us to defer initializing the
runtime until after the arguments have been parsed and we can confirm
which subcommand has been requested.

While we're at it, replace a few unwraps with errors.

wfchandler commented:
Manually validated runtime with the following patch:

```patch
diff --git a/cli/src/cli_builder.rs b/cli/src/cli_builder.rs
index 602ae05..d5ba131 100644
--- a/cli/src/cli_builder.rs
+++ b/cli/src/cli_builder.rs
@@ -298,6 +298,7 @@ impl<'a> NewCli<'a> {
                 .enable_all()
                 .build()?
         };
+        dbg!(&rt);
         rt.block_on(async move { cmd.run_cmd(sm, &ctx).await })
     }
 ```

```
$ ./target/debug/oxide version
[cli/src/cli_builder.rs:301:9] &rt = Runtime {
    scheduler: CurrentThread(
        CurrentThread,
    ),
    handle: Handle {
        inner: CurrentThread(
            current_thread::Handle { ... },
        ),
    },
    blocking_pool: BlockingPool,
}
Oxide CLI 0.9.0+20241204.0.0
Built from commit: e31ef2f63588bdc0a53be1d137151d5aad8eb6a6 (dirty)
Oxide API: 20250212.0.0

$ ./target/debug/oxide disk import ...
[cli/src/cli_builder.rs:301:9] &rt = Runtime {
    scheduler: MultiThread(
        MultiThread,
    ),
    handle: Handle {
        inner: MultiThread(
            multi_thread::Handle { ... },
        ),
    },
    blocking_pool: BlockingPool,
}
```

ahl commented on cli/src/cli_builder.rs:
```diff
@@ -269,13 +276,29 @@ impl<'a> NewCli<'a> {
+        let Some(cmd) = node.cmd.as_ref() else {
+            bail!("action not found for subcommand");
+        };
+        let rt = if cmd.as_ref().type_id() == TypeId::of::<CustomCmd<CmdDiskImport>>() {
```
I'd prefer that we make a new runtime in the disk upload subcommand or make threads manually as we deem appropriate

wfchandler replied:
I went with just using the `current_thread` executor everywhere based on our discussion.

wfchandler committed:
Just use current_thread everywhere

The differences in timing for disk imports are not meaningful for users,
so just use the `current_thread` executor everywhere.

ahl commented on cli/src/cli_builder.rs:
```diff
@@ -269,13 +269,19 @@ impl<'a> NewCli<'a> {
         let mut node = &runner;
         let mut sm = &matches;
         while let Some((sub_name, sub_matches)) = sm.subcommand() {
-            node = node.children.get(sub_name).unwrap();
```
Is this an improvement? Literally this would be the violation of a program invariant. I don't believe it's testable

wfchandler replied:
Fair, replaced with `expect`.

ahl commented on cli/src/cli_builder.rs:
```diff
@@ -269,13 +269,19 @@ impl<'a> NewCli<'a> {
         }
-        node.cmd.as_ref().unwrap().run_cmd(sm, &ctx).await
+        let Some(cmd) = node.cmd.as_ref() else {
+            bail!("action not found for subcommand");
```
As above

wfchandler replied:
👍

ahl commented on cli/src/main.rs:
```diff
@@ -96,16 +96,11 @@ pub fn make_cli() -> NewCli<'static> {
 async fn main() {
     let new_cli = make_cli();
 
-    // Spawn a task so we get this potentially chunky Future off the
```
In the past things failed on windows ci. Just FYI. If it's working now, all good

wfchandler replied:
As a sanity check I ran `cargo test` 25 times on my Windows 11 machine, no issues.

ahl commented on cli/src/cli_builder.rs:
```diff
@@ -227,7 +227,7 @@ impl<'a> NewCli<'a> {
             insecure,
             timeout,
-        } = OxideCli::from_arg_matches(&matches).unwrap();
+        } = OxideCli::from_arg_matches(&matches)?;
```
Is this testable? If not let's leave it an unwrap or replace with expect or whatever

wfchandler replied:
👍

wfchandler committed:
Replace errors with expect

We're in a bad state at these points, so we should panic. Convert the
original `unwrap` calls to `expect` so we can provide a bit more context
to users without requiring them to run the CLI again with
`RUST_BACKTRACE=1`.

ahl reviewed (approved):
looks great
